trees/cart_classifier.go

```diff
@@ -0,0 +1,442 @@
+	nInstances := len(y)
+	entropy := 0.0
+	var maxLabel int64 = 0
+
+	labelCount := convertToMap(y, labels)
+	for _, label := range labels {
+		if labelCount[label] > labelCount[maxLabel] {
+			maxLabel = label
+		}
+		p := float64(labelCount[label]) / float64(nInstances)
+		logP := math.Log2(p)
+		if p == 0 {
+			logP = 0
+		}
+		entropy += (-p * logP)
+	}
+	return entropy, maxLabel
+}
+
+func calculateClassificationLoss(y []int64, labels []int64, criterion string) (float64, int64, error) {
+	if len(y) == 0 {
+		return 0, 0, errors.New("Need atleast 1 value to compute impurity")
+	}
+	if criterion == GINI {
+		loss, modeLabel := computeGiniImpurityAndModeLabel(y, labels)
+		return loss, modeLabel, nil
+	} else if criterion == ENTROPY {
+		loss, modeLabel := computeEntropyAndModeLabel(y, labels)
+		return loss, modeLabel, nil
+	} else {
+		return 0, 0, errors.New("Invalid impurity function, choose from GINI or ENTROPY")
+	}
+}
+
+// Split the data into left node and right node based on feature and threshold
+func classifierCreateSplit(data [][]float64, feature int64, y []int64, threshold float64) ([][]float64, [][]float64, []int64, []int64) {
+	var left [][]float64
+	var right [][]float64
+	var lefty []int64
+	var righty []int64
+
+	for i := range data {
+		example := data[i]
+		if example[feature] < threshold {
+			left = append(left, example)
+			lefty = append(lefty, y[i])
+		} else {
+			right = append(right, example)
+			righty = append(righty, y[i])
+		}
+	}
+
+	return left, right, lefty, righty
+}
+
+// Function to Create New Decision Tree Classifier.
+// It assigns all of the hyperparameters by user into the tree attributes.
+func NewDecisionTreeClassifier(criterion string, maxDepth int64, labels []int64) *CARTDecisionTreeClassifier {
+	var tree CARTDecisionTreeClassifier
+	tree.criterion = strings.ToLower(criterion)
+	tree.maxDepth = maxDepth
+	tree.labels = labels
+
+	return &tree
+}
+
+// Reorder the data by feature being considered. Optimizes code by reducing the number of times we have to loop over data for splitting
+func classifierReOrderData(featureVal []float64, data [][]float64, y []int64) ([][]float64, []int64) {
+	s := NewSlice(featureVal)
+	sort.Sort(s)
+
+	indexes := s.Idx
+
+	var dataSorted [][]float64
+	var ySorted []int64
+
+	for _, index := range indexes {
+		dataSorted = append(dataSorted, data[index])
+		ySorted = append(ySorted, y[index])
+	}
+
+	return dataSorted, ySorted
+}
+
+// Update the left and right side of the split based on the threshold.
+func classifierUpdateSplit(left [][]float64, leftY []int64, right [][]float64, rightY []int64, feature int64, threshold float64) ([][]float64, []int64, [][]float64, []int64) {
+
+	for right[0][feature] < threshold {
+		left = append(left, right[0])
+		right = right[1:]
+		leftY = append(leftY, rightY[0])
+		rightY = rightY[1:]
+	}
+
+	return left, leftY, right, rightY
+}
+
+// Fit - Creates an Emppty Root Node2
```

Comment:
Nit: empty

Reply:
I might have accidentally committed trees/tmp. Should I update the same, or is there a way to exclude a file from the pull request.

Should I also modify the pull request by fixing the typo?

Reply:
As it's quite a small file (seems to be a test artefact or something), feel free to just push another commit that'll remove it. If you'd prefer, I can push a small fixup before merging.

Reply:
I pushed another commit fixing the tmp file, this typo, and the extra new line.